sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DatetimeFormatterSuite.scala

```diff
@@ -51,4 +79,36 @@ trait DatetimeFormatterSuite extends SparkFunSuite with SQLHelper with Matchers
       pattern => intercept[SparkUpgradeException](checkFormatterCreation(pattern, true))
     }
   }
+
+  test("SPARK-31939: Fix Parsing day of year when year field pattern is missing") {
+    // resolved to queryable LocaleDate or fail directly
+    assertEqual("yyyy-dd-DD", "2020-29-60", date(2020, 2, 29))
+    assertError("yyyy-dd-DD", "2020-02-60",
+      "Field DayOfMonth 29 differs from DayOfMonth 2 derived from 2020-02-29")
```

Comment:
shall we remove `derived from 2020-02-29`? looks confusing.

Reply:
what's the error message if we run the jdk time lib directly?

Reply:
> shall we remove `derived from 2020-02-29`? looks confusing.
remove it from the test cases here or the original error? It's ok to remove it here, but cannot be done from the original error because it is thrown by java time lib.

> what's the error message if we run the jdk time lib directly?

`Text '2020-02-60' could not be parsed: Conflict found: Field DayOfMonth 29 differs from DayOfMonth 2 derived from 2020-02-29` is thrown by JDK. The DatetimeException we defined just omits the leading `Text '2020-02-60' could not be parsed: `

Reply:
OK then it's fine.